src/text.c

```diff
@@ -0,0 +1,396 @@
+/* sdb - MIT - Copyright 2020 - thestr4ng3r */
+
+#include "sdb.h"
+
+#include <fcntl.h>
+#include <limits.h>
+#include <sys/stat.h>
+#if USE_MMAN
+#include <sys/mman.h>
+#endif
+
+/**
+ * ********************
+ * Plaintext SDB Format
+ * ********************
+ *
+ * Files are UTF-8 and use '\n' line endings. Always.
+ * 
+ * Lines starting with '/' denote the path of the namespace for the following data:
+ *
+ *   /some/namespace
+ * 
+ * The default path is root, just a slash also means root.
+ * These paths are always absolute from the root. Characters that must be escaped in a path are: '/', '\\', '\n', '\r':
+ *
+ *   /s\/ome/name\nspa\\ce
+ *
+ * SDB entries are written each as a single k=v line:
+ *
+ *   somekey=somevalue
+ *
+ * To distinguish these from path lines, if there is a leading '/' in the key, it must be escaped
+ * (slashes later in the line don't have to be escaped):
+ *
+ *   \/slashedkey=somevalue
+ *
+ * Other than that, at any postion, '\\', '\n' and '\r' must be escaped:
+ *   
+ *   some\\key=some\nvalue
+ *
+ * In the key, '=' must also be escaped (not necessary in the value):
+ *
+ *   some\=key=some=value
+ * 
+ * --------
+ * Example:
+ *
+ *   /
+ *   key=intheroot
+ *   \/slashedkey=somevalue
+ *   some\\key=some\nvalue
+ *   some\=key=some=value
+ *
+ *   /subns
+ *   some=stuff in the sub-namespace
+ *
+ *   /subns/deeper
+ *   this=is in /subns/deeper
+ *
+*/
+
+static int cmp_ns(const void *a, const void *b) {
+	const SdbNs *nsa = a;
+	const SdbNs *cia = b;
+	return strcmp (nsa->name, cia->name);
+}
+
+// n = position we are currently looking at
+// p = position until we have already written everything
+// flush a block of text that doesn't have to be escaped
+#define FLUSH do { if (p != n) { write (fd, p, n - p); p = n; } } while (0)
+// write and escape a string from str to fd
+#define ESCAPE_LOOP(fd, str, escapes) do { \
+		const char *p = str; \
+		const char *n = p; \
+		while (*n) { \
+			switch (*n) { escapes } \
+			n++; \
+		} \
+		FLUSH; \
+	} while (0)
+#define ESCAPE(c, repl, replsz) \
+		case c: \
+			FLUSH; \
+			p++; \
+			write (fd, "\\"repl, replsz + 1); \
+			break;
+
+static void write_path(int fd, SdbList *path) {
+	write (fd, "/", 1); // always print a /, even if path is empty
+	SdbListIter *it;
+	const char *path_token;
+	bool first = true;
+	ls_foreach (path, it, path_token) {
+		if (first) {
+			first = false;
+		} else {
+			write (fd, "/", 1);
+		}
+		ESCAPE_LOOP (fd, path_token,
+			ESCAPE ('\\', "\\", 1);
+			ESCAPE ('/', "/", 1);
+			ESCAPE ('\n', "n", 1);
+			ESCAPE ('\r', "r", 1);
+		);
+	}
+}
+
+static void write_key(int fd, const char *k) {
+	// escape leading '/'
+	if (*k == '/') {
+		write (fd, "\\", 1);
+	}
+	ESCAPE_LOOP (fd, k,
+		ESCAPE ('\\', "\\", 1);
+		ESCAPE ('=', "=", 1);
```

Comment:
Escape tab? (In all 3 functions) escaping things

Reply:
I strongly prefer not to because it's unnecessary and just uses more memory and slows down parsing.

Reply:
ping @trufae

Reply:
i dont understand how replacing \t with \\t will be a memory problem at all. but anyway, tab is a valid char for sdb, so lets keep it like this